### PR TITLE
Fix handling of mandatory meta plugins

### PR DESCRIPTION
**Pull request by jasontedor**
This commit fixes an issue with setting plugin.mandatory to include a meta-plugin. The issue here is that the names that we collect are the underlying plugins, not the meta-plugin. We should not use the underlying plugins instead using the names of non-meta plugins and the names of meta-plugins. This commit addresses this. The strategy here is that when we look at the installed plugins on the filesystem, we keep track of which ones are meta-plugins and carry this information up to where check which plugins are installed against the mandatory plugins.

Relates #28022

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/plugins/PluginsService.java [hunk 1]
```diff
@@ -281,35 +304,75 @@ public int hashCode() {
     }
 
     /**
-     * Extracts all installed plugin directories from the provided {@code rootPath} expanding meta plugins if needed.
+     * Represents a meta-plugin and the {@link Bundle}s corresponding to its constituents.
+     */
+    static class MetaBundle implements BundleCollection{
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/plugins/PluginsService.java]: nit: missing space after BundleCollection

File: server/src/main/java/org/elasticsearch/plugins/PluginsService.java [hunk 2]
```diff
@@ -135,11 +138,15 @@ public PluginsService(Settings settings, Path configPath, Path modulesDirectory,
...
-                    for (Bundle bundle : plugins) {
-                        pluginsList.add(bundle.plugin);
+                    List<BundleCollection> plugins = getPluginBundleCollections(pluginsDirectory);
+                    for (final BundleCollection plugin : plugins) {
+                        final Collection<Bundle> bundles = plugin.bundles();
+                        for (final Bundle bundle : bundles) {
+                            pluginsList.add(bundle.plugin);
+                        }
+                        seenBundles.addAll(bundles);
+                        pluginsNames.add(plugin.name());
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/plugins/PluginsService.java]: This means that a plugin inside a `meta` plugin cannot be set as mandatory ? I don't know if this is a problem or not though.

**Review by jimczi — APPROVED** [review]
LGTM, I left a minor comment and a question.

File: server/src/main/java/org/elasticsearch/plugins/PluginsService.java (hunk 2 above)
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/plugins/PluginsService.java]: You are correct. I think this is okay though, it’s limited to the user-facing plugins (what they can install and uninstall). Do you agree with that?

File: server/src/main/java/org/elasticsearch/plugins/PluginsService.java (hunk 1 above)
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/plugins/PluginsService.java]: Thanks @jimczi I will correct this.

File: server/src/main/java/org/elasticsearch/plugins/PluginsService.java (hunk 2 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/plugins/PluginsService.java]: agreed, thanks for asking.

File: server/src/main/java/org/elasticsearch/plugins/PluginsService.java (hunk 1 above)
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/plugins/PluginsService.java]: I pushed 368f5f7d8ff1885b33453da9f64900847c01d6ce.

**Comment by jasontedor** [comment]
Thanks @jimczi.

**Pull request merged**